### ISSUE TITLE
Python files with tree-sitter make NVim crash

### ISSUE THREAD

**Issue by thriveth** [nvim-treesitter/nvim-treesitter#5098]
### Describe the bug

Invoking any tree-sitter functionality in a Python file immediately makes Neovim crash. 

This may be a problem with my personal system, but I have no idea how to troubleshoot it.

### To Reproduce

1. Install nvim-treesitter using Packer, create a ~/.config/nvim/init.lua consisting only of the basic example set-up from the README, enable either tree-sitter functionality for Python (either folding, highlighting or incremental selection all seem to cause the problem on their own).

2. Open a Python file. Neovim immediately crashes and returns to terminal.

### Expected behavior

Neovim stays open and tree-sitter based functionality is enabled.

### Output of `:checkhealth nvim-treesitter`

```markdown
==============================================================================
nvim-treesitter: require("nvim-treesitter.health").check()

Installation ~
- WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
- OK `node` found v12.22.12 (only needed for :TSInstallFromGrammar)
- OK `git` executable found.
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (conda-forge gcc 13.1.0-0) 13.1.0
- OK Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

OS Info:
{
  machine = "x86_64",
  release = "6.1.0-0.deb11.7-amd64",
  sysname = "Linux",
  version = "#1 SMP PREEMPT_DYNAMIC Debian 6.1.20-2~bpo11+1 (2023-04-23)"
} ~

Parser/Features         H L F I J
  - bash                ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - latex               ✓ . ✓ . ✓
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ ✓ ✓
  - python              ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - vim                 ✓ ✓ ✓ . ✓
  - vimdoc              ✓ . . . ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang} ~
```

### Output of `nvim --version`

```text
NVIM v0.9.1
Build type: Release
LuaJIT 2.1.0-beta3

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/__w/neovim/neovim/build/nvim.AppDir/usr/share/nvim"

Run :checkhealth for more info
```

### Additional context

This is running on Debian 11. The problem persists whether I am using the Snap or the official Neovim executables, and it persists at least from version 0.8.3 and onward.

**Comment by amaanq** [comment]
Can you clone [tree-sitter-python](https://github.com/tree-sitter/tree-sitter-python) and run `cc src/scanner.c src/parser.c -fPIC -shared -Wall` ?

**Comment by clason** [comment]
Are you sure you have the latest version of the Python parser installed?

If so, can you share a Python file that crashes?

**Comment by amaanq** [comment]
I suspect it could be because of the `static_assert` in the scanner

**Comment by thriveth** [comment]
@amaanq Sorry for being a complete n00b here, but when I try to run `cc`   as you suggest, I get a "no such file or directory" error message for the initial include statement for tree_sitter/parser.h, even though the file is present and the path is correct. I am not sure what to make of that.

**Comment by thriveth** [comment]
@clason I have run `TSUPDATE` as the last thing before I filed the report. 
NVim crashes even with an empty file, as soon as it has the .py file ending.

**Comment by clason** [comment]
Clearly, that doesn't happen for most people, so we need a lot more help in nailing down what the issue is on your end. Try wiping and reinstalling Neovim and nvim-treesitter.

**Comment by amaanq** [comment]
> @amaanq Sorry for being a complete n00b here, but when I try to run `cc` as you suggest, I get a "no such file or directory" error message for the initial include statement for tree_sitter/parser.h, even though the file is present and the path is correct. I am not sure what to make of that.

I have no clue, if you're in the tree-sitter-python directory it should compile...that could maybe indicate hidden issues with how you have your environment/compiler setup

If you do get it to compile - place the output shared object into path/to/nvim-treesitter/parser/python.so

I would try @clason's suggestion first though

**Comment by thriveth** [comment]
@clason If I pass `--clean`,  Neovim cannot seem to find my installed packages, including nvim-treesitter or Packer. 

However, I can make a minimal init.lua which bootstraps packer, installs nvim-treesitter and does a minimal setup of that one, and that is all. 

Will that work, or do I need some more lua-fu to make it sufficiently minimal?

**Comment by thriveth** [comment]
@amaanq  I cloned the repo as you suggested, navigated to the directory and entered `csh`, which I never use and therefore have never changed the setup, so that should run in a standard, vanilla Debian 11 environment. Still, I get the same error, it is very weird....

**Comment by amaanq** [comment]
Does `which cc` show anything? Otherwise I can't really help, sorry

Here's my output
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/29718261/568f9065-5bf5-440d-8371-da54a3480f70)

**Comment by thriveth** [comment]
@amaanq I am quite dumbfounded, but yeah, cc is standard /usr/bin/cc, and I do exactly what you do. Even tried doing it as root, just to see if there were some weird setup hiding somewhere, but same thing. But thank you for trying anyway.

**Comment by amaanq** [comment]
Last thing I guess, does `ls -lah /usr/bin/cc` show as a symlink to gcc? 

![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/29718261/0d4fbbda-0381-417c-9009-9721b59eea98)

I also thought about it, maybe your system tries to include a system header if the include is with angled brackets - try making it quoted instead? (Or install tree-sitter systemwide via your distro if doable or locally by cloning tree-sitter/tree-sitter and running make + sudo make install)

**Comment by thriveth** [comment]
@amaanq  Replacing the angled brackets with quotations did the trick!

Now, I can try and replace the object and see if that makes a difference.

**Comment by clason** [comment]
> @clason If I pass `--clean`, Neovim cannot seem to find my installed packages, including nvim-treesitter or Packer.
> 
> However, I can make a minimal init.lua which bootstraps packer, installs nvim-treesitter and does a minimal setup of that one, and that is all, and which can be run with `nvim -u minimal.lua`.
> 
> Will that work, or do I need some more lua-fu to make it sufficiently minimal?

No, the order is important; you need both `--clean` and `-u minimal.lua` to make sure your packages are not interfering.

**Comment by thriveth** [comment]
OK, it looks like flushing out the nvim data path did the trick. Wonder if an old version of tree-sitter might have been sitting around in there causing the havoc. I thought I had gotten it down earlier, but seems like I had not. 

I apologize for the noise, but it seems to work now.

**Comment by AXGKl** [comment]
Hi, addendum, maybe useful to others:

I got the same `static_assert` error and crashing python files. Clearing the installation and starting over did not work for me.
For me, THIS was the culprint (see the `checkhealth` output in the OP):

```
(...)
- OK `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
  Version: cc (conda-forge gcc 13.1.0-0) 13.1.0
(...)
 ```

I.e. NOT the system compiler being used but the one from conda-forge (which is a binary packaging system, sth like homebrew, allows to install packages on servers w/o root perms).

And the cc binary of conda-forge seems not to play well with treesitter-python, crashing nvim...

Fix for me:

A) When I have root on the system and/or compilers are already installed:
- If not present, install the compilers with the distro's packaging system (requiring root)
- Then make sure `$PATH` is using `/usr/bin` BEFORE any conda environment.
I do this in a `vi` function in `.bashrc` or `.zshrc` - but there are other ways, e.g. a `vi` wrapper in `~/.local/bin/vi`

B) When I don't have root and no compilers present: Use `zig` as compiler.

That compiler *does* work well - better than conda's gcc binaries. 

`conda install zig` and then

`export CC="zig cc"` and 
`export CXX="zig c++"`

... before starting nvim the first time or after clearing `~/.local/share/nvim`

Note: I tried first to export `CFLAGS` to force c99 and such but w/o luck - only using `zig` did really work in a conda environment.
